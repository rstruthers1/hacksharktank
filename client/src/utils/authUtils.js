// authUtils.js
export const isUserLoggedIn = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }

    // Check if expired, else return true
    try {
        const {exp} = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now < expiration;
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }

};

export const getLoggedInUser = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return null;
    }

    // Check if expired, else return true
    try {

        const {id, username, email} = JSON.parse(atob(token.split('.')[1])); // Decode payload
        console.log(`id: ${id}, username: ${username}, email: ${email}`)
        return {id, email};
    } catch {
        return null; // if there's an error decoding the token, consider it invalid
    }
}

export const isUserAdmin = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }
    // Check if expired
    let user = null;
    try {
        user = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(user.exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now < expiration && user.roles.includes('admin');
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }
}

// Returns true if token exists and is expired, else false
export const isSessionExpired = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }

    // Check if expired, else return true
    try {
        const {exp} = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now > expiration;
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }
}

export const logoutUser = () => {
    localStorage.removeItem('token');
}
