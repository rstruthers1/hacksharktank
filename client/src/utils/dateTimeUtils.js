const MILLIS_IN_ONE_DAY = 86400000;

export const getTodayAtMidnight = () => {
    const today = new Date();
    today.setHours(0,0,0,0);
    return today;
}

export const getMinEndDate = (startDate) => {
    if (!startDate) {
        return getTodayAtMidnight().getTime() + MILLIS_IN_ONE_DAY;
    }
    try {
        const startDateAsDate = new Date(startDate);
        const dayAfter = new Date(startDateAsDate.getTime() + MILLIS_IN_ONE_DAY);
        return dayAfter.getTime();
    } catch (err) {
        console.error(err);
        return getTodayAtMidnight().getTime() + MILLIS_IN_ONE_DAY;
    }
}

export const formatDateTime = (dateTime) => {
    if (!dateTime) {
        return '';
    }
    try {
        const dateTimeAsDate = new Date(dateTime);
        return dateTimeAsDate.toLocaleString();
    } catch (err) {
        console.error(err);
        return '';
    }
}

export const formatDateTimeAsDate  = (date) => {
    if (!date) {
        return '';
    }
    try {
        const dateAsDate = new Date(date);
        return dateAsDate.toLocaleDateString();
    } catch (err) {
        console.error(err);
        return '';
    }
}