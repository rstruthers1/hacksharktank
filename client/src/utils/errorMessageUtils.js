import React from 'react';

export const getErrorMessage = (error) => {
    if (!error?.data?.message) {
        return 'Unknown error';
    }
    if (Array.isArray(error.data.message)) {
        const errorMessages = error.data.message;
        return (
            <ul>
                {errorMessages.map(errorMessage => {
                    return (
                        <li>{errorMessage.msg}</li>
                    );
                })}
            </ul>
        );
    } else {
        return error.data.message;
    }
}