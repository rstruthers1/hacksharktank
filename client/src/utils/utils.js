// utils.js
export const isUserLoggedIn = () => {
    const token = localStorage.getItem('token');

    if (!token) {
        return false;
    }

    // Optionally check for token expiry if your JWT contains an expiry claim
    try {
        const { exp } = JSON.parse(atob(token.split('.')[1])); // Decode payload
        const expiration = new Date(exp * 1000); // Convert to milliseconds
        const now = new Date();
        return now < expiration;
    } catch {
        return false; // if there's an error decoding the token, consider it invalid
    }
};
