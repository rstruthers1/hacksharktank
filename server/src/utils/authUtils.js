const knex = require("../config/knex");

const userIsSiteAdminOrHackathonMember = async (req, hackathonId) => {
    const authUserRoles = req.user.roles;
    let isSiteAdminOrHackathonMember = false
    if (authUserRoles && authUserRoles.includes('admin')) {
        isSiteAdminOrHackathonMember = true;
    } else {
        // check to see if logged-in user is in hackathon
        const hackathonUser = await knex('hackathon_user')
            .where('hackathonId', hackathonId)
            .where('userId', req.user.id)
            .first();
        if (hackathonUser) {
            isSiteAdminOrHackathonMember = true;
        }
    }
    return isSiteAdminOrHackathonMember;
}

const userIsSiteAdminOrHasHackathonRole = async (req, hackathonId, roles) => {
    const authUserRoles = req.user.roles;
    let isSiteAdminOrHasHackathonRole = false
    if (authUserRoles && authUserRoles.includes('admin')) {
        isSiteAdminOrHasHackathonRole = true;
    } else {
        // check to see if logged-in user is in hackathon
        const hackathonUserRoles = await knex('hackathon_user')
            .join('hackathon_user_role', 'hackathon_user.id', '=', 'hackathon_user_role.hackathonUserId')
            .join('hackathon_role', 'hackathon_user_role.hackathonRoleId', '=', 'hackathon_role.id')
            .where('hackathonId', hackathonId)
            .where('userId', req.user.id)
            .select('hackathon_role.name as role');
        if (hackathonUserRoles && roles && roles.length > 0 && hackathonUserRoles.some(hackathonUserRole => roles.includes(hackathonUserRole.role))) {
            isSiteAdminOrHasHackathonRole = true;
        }
    }
    return isSiteAdminOrHasHackathonRole;
}

// make visible to other files
module.exports = {
    userIsSiteAdminOrHackathonMember,
    userIsSiteAdminOrHasHackathonRole
}