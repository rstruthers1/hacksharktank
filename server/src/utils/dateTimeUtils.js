const moment = require('moment');


const convertDateToDBFormat = (date) => {
    return moment(date).format('YYYY-MM-DD HH:mm:ss');
};

const getTodayAtMidnight = () => {
    const today = new Date();
    today.setHours(0,0,0,0);
    return today;
}

module.exports = {
    convertDateToDBFormat,
    getTodayAtMidnight
};