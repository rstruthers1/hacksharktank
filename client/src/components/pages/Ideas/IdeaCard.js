import React, {useState} from 'react';
import {Card} from "react-bootstrap";
import {MdDelete, MdEdit} from 'react-icons/md';
import {getLoggedInUser} from "../../../utils/authUtils";
import ConfirmModalDialog from "../../common/ConfirmModalDialog";
import {useDeleteHackathonIdeaMutation} from "../../../apis/hackathonIdeaApi";
import {toast} from "react-toastify";

const IdeaCard = ({idea}) => {
    const [showDeleteModal, setShowDeleteModal] = useState(false);
    const [deleteHackathonIdea, {error: deleteHackathonIdeaError}] = useDeleteHackathonIdeaMutation();
    const [ideaToDelete, setIdeaToDelete] = useState(null);

    const isDeleteDisabled = () => {
        return !getLoggedInUser() || getLoggedInUser().id !== idea.userId;
    }

    const deleteIconStyle = {
        cursor: isDeleteDisabled() ? 'default' : 'pointer',
        color: isDeleteDisabled() ? '#CCCCCC' : 'black', // Grey color for disabled state
    };

    const deleteConfirmed = async () => {
        // call api to delete the idea
        await deleteHackathonIdea({hackathonId: idea.hackathonId, ideaId: idea.id}).unwrap()
            .then(() => {
                toast.success(`Successfully deleted idea "${ideaToDelete?.title}"`, {
                        position: toast.POSITION.TOP_CENTER
                    }
                );
            })
            .catch((err) => {
                toast.error(`Error deleting idea "${ideaToDelete?.title}"`, {
                        position: toast.POSITION.TOP_CENTER
                    }
                );
            });
    }

    const handleDeleteClicked = (ev) => {
        ev.preventDefault();
        if (isDeleteDisabled()) return;
        setIdeaToDelete({...idea})
        setShowDeleteModal(true);
    }

    const handleEditClicked = (ev) => {
        ev.preventDefault();

    }

    return (
        <>
            <Card className="mb-3">
                <Card.Header>
                    {idea.title}
                    <span style={{float: 'right'}}> {/* Container for the icons */}
                        <MdEdit
                            style={{cursor: 'pointer', marginRight: '5px'}} // Style for edit icon
                            onClick={handleEditClicked}
                        />
                        <MdDelete
                            style={deleteIconStyle} // Style for delete icon
                            onClick={handleDeleteClicked}
                            data-testid="delete-icon"
                        />
                    </span>
                </Card.Header>
                <Card.Body>
                    <Card.Text dangerouslySetInnerHTML={{__html: idea.description}}/>
                </Card.Body>
            </Card>
            <ConfirmModalDialog
                show={showDeleteModal}
                title="Delete Idea"
                message={`Are you sure you want to delete this idea: "${idea.title}?"`}
                onConfirm={() => {
                    setShowDeleteModal(false);
                    deleteConfirmed()
                }}
                confirmLabel='Delete'
                onCancel={() => setShowDeleteModal(false)}/>
        </>
    );
}

export default IdeaCard;
