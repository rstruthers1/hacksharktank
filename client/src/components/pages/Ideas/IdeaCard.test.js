import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import IdeaCard from './IdeaCard'; // Adjust the import path as necessary
import { getLoggedInUser } from '../../../utils/authUtils';
import renderer from "react-test-renderer"; // Adjust the import path as necessary

// Mocking necessary modules and functions
jest.mock('../../../utils/authUtils');
jest.mock('../../../apis/hackathonIdeaApi', () => ({
    useDeleteHackathonIdeaMutation: () => [jest.fn(), {error: null}]
}));
jest.mock('react-toastify', () => ({
    toast: {
        success: jest.fn(),
        error: jest.fn(),
        POSITION: {
            TOP_CENTER: 'top-center'
        }
    }
}));

describe('IdeaCard Component', () => {
    const ideaMock = {
        id: '1',
        title: 'Test Idea',
        description: '<p>Test Description</p>',
        userId: 'user123',
        hackathonId: 'hackathon1'
    };

    beforeEach(() => {
        // Setup or reset the mocks before each test
        getLoggedInUser.mockReturnValue({ id: 'user123' }); // Mock the logged in user
    });

    it('should render correctly', () => {
        const tree = renderer.create(<IdeaCard idea={ideaMock}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders the component correctly', () => {
        render(<IdeaCard idea={ideaMock} />);
        expect(screen.getByText('Test Idea')).toBeInTheDocument();
        expect(screen.getByText('Test Description')).toBeInTheDocument();
    });

    it('shows delete icon with correct style for authorized user', () => {
        render(<IdeaCard idea={ideaMock} />);
        const deleteIcon = screen.getByTestId('delete-icon'); // Add data-testid="delete-icon" to MdDelete in your component
        expect(deleteIcon).toBeInTheDocument();
        expect(deleteIcon).toHaveStyle('cursor: pointer');
    });

    it('delete icon click opens confirmation modal', () => {
        render(<IdeaCard idea={ideaMock} />);
        const deleteIcon = screen.getByTestId('delete-icon');
        fireEvent.click(deleteIcon);
        expect(screen.getByText(/Are you sure you want to delete this idea/i)).toBeInTheDocument();
    });

    // Add more tests as needed, like for modal confirmation, cancellation, disabled delete icon, etc.
});
