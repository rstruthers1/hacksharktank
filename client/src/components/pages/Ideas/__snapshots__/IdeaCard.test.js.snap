// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeaCard Component should render correctly 1`] = `
<div
  className="mb-3 card"
>
  <div
    className="card-header"
  >
    Test Idea
    <span
      style={
        Object {
          "float": "right",
        }
      }
    >
       
      <svg
        fill="currentColor"
        height="1em"
        onClick={[Function]}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
            "cursor": "pointer",
            "marginRight": "5px",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
      <svg
        data-testid="delete-icon"
        fill="currentColor"
        height="1em"
        onClick={[Function]}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": "black",
            "cursor": "pointer",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
  </div>
  <div
    className="card-body"
  >
    <p
      className="card-text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Test Description</p>",
        }
      }
    />
  </div>
</div>
`;
